File: backend/hpo/message_queue.py
from collections import deque
from config import MAX_QUEUE_DEPTH

# Per-user outbound queues
_queues: dict[str, deque] = {}


def enqueue_message(user_id: str, packet_data: dict):
    if user_id not in _queues:
        _queues[user_id] = deque(maxlen=MAX_QUEUE_DEPTH)
    _queues[user_id].append(packet_data)


def dequeue_message(user_id: str) -> dict | None:
    q = _queues.get(user_id)
    if q and len(q) > 0:
        return q.popleft()
    return None


def get_queue_depth(user_id: str) -> int:
    q = _queues.get(user_id)
    return len(q) if q else 0


def clear_queue(user_id: str):
    _queues.pop(user_id, None)

File: server/message-queue.ts
import { EncryptedPacket } from "../src/socket/events";
import { MAX_QUEUE_DEPTH } from "./config";

export interface HPOPacketData {
  type: "real" | "cover";
  packet: EncryptedPacket | null;
  recipientId: string;
}

// Per-user outbound queue
const queues = new Map<string, HPOPacketData[]>();

export function enqueueMessage(userId: string, packet: HPOPacketData): void {
  if (!queues.has(userId)) {
    queues.set(userId, []);
  }

  const queue = queues.get(userId)!;

  // Drop oldest if queue is full
  if (queue.length >= MAX_QUEUE_DEPTH) {
    queue.shift();
  }

  queue.push(packet);
}

export function dequeueMessage(userId: string): HPOPacketData | null {
  const queue = queues.get(userId);
  if (!queue || queue.length === 0) return null;
  return queue.shift()!;
}

export function getQueueDepth(userId: string): number {
  return queues.get(userId)?.length ?? 0;
}

export function clearQueue(userId: string): void {
  queues.delete(userId);
}
